File: app/src/main/java/com/example/pam_project/features/tasks/list/TaskActivity.java
package com.example.pam_project.features.tasks.list;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.pam_project.R;
import com.example.pam_project.di.ApplicationContainer;
import com.example.pam_project.di.ApplicationContainerLocator;
import com.example.pam_project.features.lists.edit.EditListActivity;
import com.example.pam_project.features.lists.list.OnListClickedListener;
import com.example.pam_project.features.tasks.create.CreateTaskActivity;
import com.example.pam_project.repositories.lists.ListsRepository;
import com.example.pam_project.repositories.tasks.TaskRepository;
import com.example.pam_project.utils.ActivityRegistry;
import com.example.pam_project.utils.ActivityResultCode;
import com.example.pam_project.utils.TaskStatus;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;

import java.util.List;
import java.util.Objects;

public class TaskActivity extends AppCompatActivity implements TaskView, OnListClickedListener {
    private RecyclerView recyclerView;
    private TaskAdapter adapter;
    private TaskPresenter taskPresenter;
    private long listId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        String id = getIntent().getData().getQueryParameter("id");
        this.listId = Long.parseLong(id);

        final ApplicationContainer container = ApplicationContainerLocator
                .locateComponent(this);

        final TaskRepository taskRepository = container.getTasksRepository();
        final ListsRepository listsRepository = container.getListsRepository();

        setContentView(R.layout.activity_task);
        setup();
        taskPresenter = new TaskPresenter(taskRepository, listsRepository, this, listId);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.edit_action_bar, menu);
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        taskPresenter.onViewAttached();
    }

    private void setup() {
        recyclerView = findViewById(R.id.allTasks);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));

        final ItemTouchHelper touchHelper = new ItemTouchHelper(setSwippableItems());
        touchHelper.attachToRecyclerView(recyclerView);
        setExtendedFloatingButtonAction();
    }

    private void setExtendedFloatingButtonAction() {
        ExtendedFloatingActionButton addTaskFAB = findViewById(R.id.extended_fab_add_task);
        addTaskFAB.setOnClickListener(view -> taskPresenter.onButtonAddClicked());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == ActivityRegistry.CREATE_TASK_ACTIVITY.ordinal()) {
            if(resultCode == ActivityResultCode.CREATE_LIST_CODE.ordinal()){
                // Code for create task
            }
        } else if(requestCode == ActivityRegistry.EDIT_LIST_ACTIVITY.ordinal()) {
            if (resultCode == ActivityResultCode.EDIT_LIST_CODE.ordinal()) {
                // Code for edit list
            } else if (resultCode == ActivityResultCode.DELETE_LIST_CODE.ordinal()) {
                finish();
            }
        } else if(requestCode == ActivityRegistry.EDIT_TASK_ACTIVITY.ordinal()){
            if (resultCode == ActivityResultCode.EDIT_TASK_CODE.ordinal()) {
                // Code for edit task
            } else if (resultCode == ActivityResultCode.DELETE_LIST_CODE.ordinal()) {
                // Code for delete task
            }
        }
    }

    @Override
    public void showTasks() {
        adapter = new TaskAdapter();
        adapter.setOnClickedListener(this);
        recyclerView.setAdapter(adapter);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        long id = item.getItemId();

        if (id == R.id.edit_list_button) {
            taskPresenter.onEditList();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void bindTasks(List<TaskInformation> model) {
        adapter.update(model);
        taskPresenter.onEmptyTask();
    }

    @Override
    public void bindHeaders(int[] headers) {
        String[] titleHeaders = {getString(headers[0]), getString(headers[1])};
        recyclerView.addItemDecoration(new CustomItemDecorator(this, R.layout.text_header, titleHeaders));
    }

    @Override
    public void showEmptyMessage() {
        TextView text = findViewById(R.id.empty_list_message);
        View emptyDataMessage = findViewById(R.id.empty_task);

        if(adapter.getItemCount() == 0) {
            text.setText(R.string.empty_task_message);
            emptyDataMessage.setVisibility(View.VISIBLE);
        } else {
            emptyDataMessage.setVisibility(View.GONE);
        }
    }

    @Override
    public void showEditList() {
        Intent activityIntent = new Intent(getApplicationContext(), EditListActivity.class);
        activityIntent.putExtra("id", listId);
        startActivityForResult(activityIntent, ActivityRegistry.EDIT_LIST_ACTIVITY.ordinal());
    }

    @Override
    public void bindTask(TaskInformation model) {
        adapter.addItem(model);
    }

    @Override
    public void bindListName(final String listName) {
        Objects.requireNonNull(getSupportActionBar()).setTitle(listName);
    }

    @Override
    public void showAddTask() {
        Intent activityIntent = new Intent(getApplicationContext(), CreateTaskActivity.class);
        activityIntent.putExtra("id", listId);
        startActivityForResult(activityIntent, ActivityRegistry.CREATE_TASK_ACTIVITY.ordinal());
    }

    @Override
    public void showTaskContent(final long id) {
        String uri = "pam://edit/task?id=";
        Intent activityIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri + id));
        startActivityForResult(activityIntent, ActivityRegistry.EDIT_TASK_ACTIVITY.ordinal());
    }

    @Override
    public void onTaskStatusEdit(final TaskInformation model, final int adapterPosition) {
        adapter.update(model, adapterPosition);
        if(adapter.areAllComplete())
            finish();
    }

    private ItemTouchHelper.SimpleCallback setSwippableItems() {
        return new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {

            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView1,
                                  @NonNull RecyclerView.ViewHolder dragged,
                                  @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder swippedItem, int direction) {
                TaskInformation taskInformation = adapter.getItem(swippedItem.getAdapterPosition());
                taskPresenter.onTaskChange(taskInformation, swippedItem.getAdapterPosition());
            }

            @Override
            public boolean isItemViewSwipeEnabled() {
                return true;
            }

            @Override
            public int getMovementFlags(@NonNull RecyclerView recyclerView,
                                        @NonNull RecyclerView.ViewHolder swippedItem) {

                final int dragFlags = 0;
                final int leftSwipeFlag = ItemTouchHelper.LEFT;
                final int rightSwipeFlag = ItemTouchHelper.RIGHT;

                TaskInformation taskInformation = adapter.getItem(swippedItem.getAdapterPosition());
                if (taskInformation.getStatus().equals(TaskStatus.DONE))
                    return makeMovementFlags(dragFlags, rightSwipeFlag);

                return makeMovementFlags(dragFlags, leftSwipeFlag);
            }
        };
    }

    @Override
    protected void onStop() {
        super.onStop();
        taskPresenter.onViewDetached();
    }

    @Override
    public void onClick(final long id) {
        taskPresenter.onTaskClicked(id);
    }
}

File: app/src/main/java/com/example/pam_project/features/tasks/list/TaskActivity.kt
package com.example.pam_project.features.tasks.list

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.pam_project.R
import com.example.pam_project.di.ApplicationContainerLocator
import com.example.pam_project.features.lists.edit.EditListActivity
import com.example.pam_project.features.lists.list.OnListClickedListener
import com.example.pam_project.features.tasks.create.CreateTaskActivity
import com.example.pam_project.utils.constants.ActivityResultCode
import com.example.pam_project.utils.constants.TaskStatus
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton

class TaskActivity : AppCompatActivity(), TaskView, OnListClickedListener {
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: TaskAdapter
    private lateinit var presenter: TaskPresenter
    private var listId: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_task)
        val id = intent.data!!.getQueryParameter("id")
        listId = id!!.toLong()
        setUpView()
        createPresenter()
    }

    private fun createPresenter() {
        val container = ApplicationContainerLocator.locateComponent(this)
        val schedulerProvider = container?.schedulerProvider
        val taskRepository = container?.tasksRepository
        val listsRepository = container?.listsRepository
        presenter = TaskPresenter(schedulerProvider, taskRepository, listsRepository,
                this, listId)
    }

    private fun setUpView() {
        recyclerView = findViewById(R.id.allTasks)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        val touchHelper = ItemTouchHelper(setSwippableItems())
        touchHelper.attachToRecyclerView(recyclerView)
        setExtendedFloatingButtonAction()
        adapter = TaskAdapter()
        adapter.setOnClickedListener(this)
        recyclerView.adapter = adapter
    }

    override fun onStart() {
        super.onStart()
        presenter.onViewAttached()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.edit_action_bar, menu)
        return true
    }

    private fun setExtendedFloatingButtonAction() {
        val addTaskFAB = findViewById<ExtendedFloatingActionButton>(R.id.extended_fab_add_task)
        addTaskFAB.setOnClickListener { presenter.onButtonAddClicked() }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == EDIT_LIST_ACTIVITY) {
            if (resultCode == ActivityResultCode.DELETE_LIST_CODE.ordinal) {
                finish()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId.toLong()
        if (id == R.id.edit_list_button.toLong()) {
            presenter.onEditList()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun bindTasks(model: List<TaskInformation?>?) {
        adapter.update(model)
        presenter.onEmptyTask()
    }

    override fun bindHeaders(headers: IntArray) {
        val titleHeaders = arrayOf(getString(headers[0]), getString(headers[1]))
        recyclerView.addItemDecoration(CustomItemDecorator(this, R.layout.text_header, titleHeaders))
    }

    override fun showEmptyMessage() {
        val text = findViewById<TextView>(R.id.empty_list_message)
        val emptyDataMessage = findViewById<View>(R.id.empty_task)
        if (adapter.itemCount == 0) {
            text.setText(R.string.empty_task_message)
            emptyDataMessage.visibility = View.VISIBLE
        } else {
            emptyDataMessage.visibility = View.GONE
        }
    }

    override fun showEditList() {
        val activityIntent = Intent(applicationContext, EditListActivity::class.java)
        activityIntent.putExtra("id", listId)
        startActivityForResult(activityIntent, EDIT_LIST_ACTIVITY)
    }

    override fun onTasksReceivedError() {
        Toast.makeText(applicationContext, getString(R.string.error_tasks_fetch), Toast.LENGTH_LONG).show()
    }

    override fun onTaskUpdatedError() {
        Toast.makeText(applicationContext, getString(R.string.error_task_update), Toast.LENGTH_LONG).show()
    }

    override fun bindListName(name: String?) {
        supportActionBar?.title = name
    }

    override fun showAddTask() {
        val activityIntent = Intent(applicationContext, CreateTaskActivity::class.java)
        activityIntent.putExtra("id", listId)
        startActivity(activityIntent)
    }

    override fun showTaskContent(id: Long) {
        val uri = "pam://edit/task?id="
        val activityIntent = Intent(Intent.ACTION_VIEW, Uri.parse(uri + id))
        startActivity(activityIntent)
    }

    override fun onTaskStatusEdit(model: TaskInformation?, adapterPosition: Int) {
        adapter.update(model, adapterPosition)
        if (adapter.areAllComplete()) finish()
    }

    private fun setSwippableItems(): ItemTouchHelper.SimpleCallback {
        return object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            override fun onMove(recyclerView1: RecyclerView,
                                dragged: RecyclerView.ViewHolder,
                                target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(swippedItem: RecyclerView.ViewHolder, direction: Int) {
                val taskInformation = adapter.getItem(swippedItem.adapterPosition)
                presenter.onTaskChange(taskInformation, swippedItem.adapterPosition)
            }

            override fun isItemViewSwipeEnabled(): Boolean {
                return true
            }

            override fun getMovementFlags(recyclerView: RecyclerView,
                                          swippedItem: RecyclerView.ViewHolder): Int {
                val dragFlags = 0
                val leftSwipeFlag = ItemTouchHelper.LEFT
                val rightSwipeFlag = ItemTouchHelper.RIGHT
                val taskInformation = adapter.getItem(swippedItem.adapterPosition)
                return if (taskInformation!!.status == TaskStatus.DONE) makeMovementFlags(dragFlags, rightSwipeFlag) else makeMovementFlags(dragFlags, leftSwipeFlag)
            }
        }
    }

    override fun onClick(id: Long) {
        presenter.onTaskClicked(id)
    }

    override fun onStop() {
        super.onStop()
        presenter.onViewDetached()
    }

    companion object {
        private const val EDIT_LIST_ACTIVITY = 1
    }
}